flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/InternalKeyContextImpl.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.heap;
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.state.KeyGroupRange;
+
+/**
+ * The default {@link InternalKeyContext} implementation.
+ *
+ * @param <K> Type of the key.
+ */
+public class InternalKeyContextImpl<K> implements InternalKeyContext<K> {
+	/** Range of key-groups for which this backend is responsible. */
+	private final KeyGroupRange keyGroupRange;
+	/** The number of key-groups aka max parallelism. */
+	private final int numberOfKeyGroups;
+
+	/** The currently active key. */
+	private K currentKey;
+	/** The key group of the currently active key. */
+	private int currentKeyGroupIndex;
+	/** {@link TypeSerializer} for the state backend key type. */
+	private TypeSerializer<K> currentKeySerializer;
```

Comment:
As explained above:
Maybe we should use the `StateSerializerProvider` here, to improve structural dependency a bit.
i.e. when initializing the key context in the builder, instantiate it with the `StateSerializerProvider`.

Reply:
As a second thought:
either with your current approach, or my suggestion, either way we'll still be leaking mutability of the key serializer in the created backend instances (they are not final).

Maybe the key context should also follow a builder pattern, where the restore operation uses setups an `InternalKeyContextBuilder`. Then, after the restore operation, when the backend is created, we build the final key context.

Reply:
After a second thought, key serializer is not required and could be final in both `StateTable` and backend. Have removed `keySerializer` field from `InternalKeyContextImpl` in the latest commit, please check and let me know whether it looks good. Thanks.